src/main/scala/io/iohk/ethereum/blockchain/sync/RegularSync.scala

```diff
@@ -12,166 +12,233 @@ import io.iohk.ethereum.network.EtcPeerManagerActor.PeerInfo
+    topOfTheChain: Boolean,
+    resolvingBranches: Boolean,
+    resumeRegularSyncTimeout: Option[Cancellable],
+    missingStateNodeRetry: Option[MissingStateNodeRetry]
+  ): Unit = {
     resumeRegularSyncTimeout.foreach(_.cancel)
-    resumeRegularSyncTimeout = None
+    context become running(waitingForActor, headersQueue, topOfTheChain, resolvingBranches, None, missingStateNodeRetry)
   }
 
-  def handleNewBlockMessages: Receive = {
+  // scalastyle:off method.length
+  def handleNewBlockMessages(
+    waitingForActor: Option[ActorRef],
+    headersQueue: Seq[BlockHeader],
+    topOfTheChain: Boolean,
+    resolvingBranches: Boolean
+  ): Receive = {
     case MessageFromPeer(NewBlock(newBlock, _), peerId) =>
       //we allow inclusion of new block only if we are not syncing
-      if (notDownloading() && topOfTheChain) {
-        log.debug(s"Handling NewBlock message for block (${newBlock.idTag})")
+      if (notDownloading(waitingForActor, headersQueue, resolvingBranches) && topOfTheChain) {
+        log.debug(s"Handling NewBlock message for block (${ newBlock.idTag })")
         val importResult = Try(ledger.importBlock(newBlock))
 
         importResult match {
-          case Success(result) => result match {
-            case BlockImportedToTop(newBlocks, newTds) =>
-              broadcastBlocks(newBlocks, newTds)
-              updateTxAndOmmerPools(newBlocks, Nil)
-              log.info(s"Added new block ${newBlock.header.number} to the top of the chain received from $peerId")
-
-            case BlockEnqueued =>
-              ommersPool ! AddOmmers(newBlock.header)
-              log.debug(s"Block ${newBlock.header.number} (${Hex.toHexString(newBlock.header.hash.toArray)}) from $peerId " +
-                s"added to queue")
-
-            case DuplicateBlock =>
-              log.debug(s"Ignoring duplicate block ${newBlock.header.number} (${Hex.toHexString(newBlock.header.hash.toArray)}) from $peerId")
-
-            case UnknownParent =>
-              // This is normal when receiving broadcasted blocks
-              log.debug(s"Ignoring orphaned block ${newBlock.header.number} (${Hex.toHexString(newBlock.header.hash.toArray)}) from $peerId")
-
-            case ChainReorganised(oldBranch, newBranch, totalDifficulties) =>
-              updateTxAndOmmerPools(newBranch, oldBranch)
-              broadcastBlocks(newBranch, totalDifficulties)
-              log.debug(s"Imported block ${newBlock.header.number} (${Hex.toHexString(newBlock.header.hash.toArray)}) from $peerId, " +
-                s"resulting in chain reorganisation: new branch of length ${newBranch.size} with head at block " +
-                s"${newBranch.last.header.number} (${Hex.toHexString(newBranch.last.header.hash.toArray)})")
-
-            case BlockImportFailed(error) =>
-              blacklist(peerId, blacklistDuration, error)
-          }
+          case Success(result) =>
+            val headerHash = hash2string(newBlock.header.hash)
```

Comment:
`headerHash` and `newNumber` could be lazy because not all cases are using both of them